src/CachedRemoteRequest.php

```diff
@@ -0,0 +1,68 @@
+<?php
+/**
+ * Class CachedRemoteRequest.
+ *
+ * @package Amp\AmpWP
+ */
+
+namespace Amp\AmpWP;
+
+use Amp\Exception\FailedToFetchFromRemoteUrl;
+use Amp\RemoteRequest;
+
+/**
+ * Caching decorator for RemoteRequest implementations.
+ *
+ * Caching uses WordPress transients.
+ *
+ * @package Amp\AmpWP
+ */
+final class CachedRemoteRequest implements RemoteRequest {
+
+
+	/**
+	 * Remote request object to decorate with caching.
+	 *
+	 * @var RemoteRequest
+	 */
+	private $remote_request;
+
+	/**
+	 * Cache expiration time in hours.
+	 *
+	 * @var int
+	 */
+	private $expiry;
+
+	/**
+	 * Instantiate a CachedRemoteRequest object.
+	 *
+	 * This is a decorator that can wrap around an existing remote request object to add a caching layer.
+	 *
+	 * @param RemoteRequest $remote_request Remote request object to decorate with caching.
+	 * @param int           $expiry         Optional. Cache expiry in hours. Defaults to 24 hours.
+	 */
+	public function __construct( RemoteRequest $remote_request, $expiry = 24 ) {
+		$this->remote_request = $remote_request;
+		$this->expiry         = $expiry;
+	}
+
+	/**
+	 * Fetch the contents of a remote request.
+	 *
+	 * @param string $url URL to fetch.
+	 * @return string Contents retrieved from the remote URL.
+	 * @throws FailedToFetchFromRemoteUrl If fetching the contents from the URL failed.
+	 */
+	public function fetch( $url ) {
+		$cache_key = 'amp_remote_request_' . md5( __CLASS__ . $url );
+		$result    = get_transient( $cache_key );
+
+		if ( false === $result ) {
+			$result = $this->remote_request->fetch( $url );
+			set_transient( $cache_key, $result, $this->expiry * HOUR_IN_SECONDS );
+		}
```

Comment:
Given that a fetch failure would be catastrophic for optimized AMP, what if this supported returning stale results? In other words, what if the transient itself had no expiration time, but the last fetch time were stored with the cached value. Then it could get the cached value and its last cached time, and then only fetch if the last cached time is greater than the expiry. If the fetch fails, the stale value could still be served but then a flag could be set to say that it failed. The next `fetch` call could then look at that flag and implement some backoff logic to keep trying.

Reply:
Yes, that would be preferable. I wanted to keep it simple here, but am happy to change for a more robust caching mechanism. I'd also add default fallback data as well then, in case the very first initial fetch fails.